jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/gzip/GzipResponse.java

```diff
@@ -249,136 +271,162 @@ protected void onCompleteFailure(Throwable x)
         @Override
         protected Action process() throws Exception
         {
-            // If we have no deflater
-            if (_deflaterEntry == null)
+            if (LOG.isDebugEnabled())
+                LOG.debug("GzipBufferCB.process(): _last={}, _buffer={}, _content={}", _last, BufferUtil.toDetailString(_buffer), BufferUtil.toDetailString(_content));
+
+            GZState gzstate = _state.get();
+
+            // Are we finished?
+            if (gzstate == GZState.FINISHED)
             {
                 // then the trailer has been generated and written below.
                 // We have finished compressing the entire content, so
                 // cleanup and succeed.
-                if (_buffer != null)
-                {
-                    getRequest().getComponents().getByteBufferPool().release(_buffer);
-                    _buffer = null;
-                }
-                if (_copy != null)
-                {
-                    getRequest().getComponents().getByteBufferPool().release(_copy);
-                    _copy = null;
-                }
+                cleanup();
                 return Action.SUCCEEDED;
             }
 
             // If we have no buffer
             if (_buffer == null)
             {
-                // allocate a buffer and add the gzip header.
                 _buffer = getRequest().getComponents().getByteBufferPool().acquire(_bufferSize, false);
-                BufferUtil.fill(_buffer, GZIP_HEADER, 0, GZIP_HEADER.length);
+                // Per RFC-1952, GZIP is LITTLE_ENDIAN
+                _buffer.order(ByteOrder.LITTLE_ENDIAN);
```

Comment:
Whilst this is true, do we ever do any non-byte operations on the buffer (e.g. adding integers or longs)?

If we start using that aspect of buffers, then we will need to reset the endian state of buffers as we recycle them through pools.

Maybe best is we just avoid non-byte operations and put some javadoc on our buffer pool saying that endianness is platform default and that any usage of a buffer that switches the endianess should revert it before putting back in the buffer.... or maybe we say that we don't guarantee the endian state of buffers from the pool?

It's a little can of worms... best just to do byte operations :)

Reply:
Yes, we do multi-byte put (of ints).
And those _must_ be in LITTLE_ENDIAN mode (per spec).

There's no thing as "platform default" for endianness in ByteBuffer, it's always BIG_ENDIAN (per [javadoc](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/nio/ByteBuffer.html))

Reply:
OK, so we need to put these buffers back to `BIG_ENDIAN` before releasing them back to the pool.  We could make the pool do it, but let's not add the expense when it is only needed here.